Add `pause` and `resume` API's to `Loop`

p4checo opened the pull request:
## Motivation

On some scenarios, it's useful to "pause" the `Loop` so that we stop processing events for some reason (e.g. to stop a `Loop` backed Service). Following the same reasoning, it becomes necessary to have a "resume" mechanism so that the Loop starts processing events again.

Given `Loop` now starts automatically and `stop` is designed as a tear down mechanism to be used on dealloc and dispose all observations, some new API's are required so that we can unplug/replug feedbacks to achieve the above mentioned pause/resume behavior.

 ⚠️ Note: Tests missing, as I wanted to validate the approach first 🙏 

## Changes

- Create new `plugFeedbacks` and `unplugFeedbacks` API's in `Floodgate`, which establish and dispose feedbacks observations, respectively. Floodgate now retains the feedbacks passed in on `bootstrap` to use them on `plugFeedbacks`.

- Add `pause` and `resume` API's to `LoopBoxBase`.

- Implement `pause` and `resume` API's in `RootLoopBox`, which unplug and replug the feedbacks on the `Floodgate`, respectively.

- Implement `pause` and `resume` API's in `ScopedLoopBox`, which forward the calls to their root, respectively.

- Fixed .podspec to include swift files from all folders.

p4checo committed:
Add `pause` and `resume` API's to `Loop`

On some scenarios, it's useful to "pause" the `Loop` so that we stop
processing events for some reason (e.g. to stop a `Loop` backed
Service). Following the same reasoning, it becomes necessary to have a
"resume" mechanism so that the Loop starts processing events again.

Given `Loop` now starts automatically and `stop` is designed as a tear
down mechanism to be used on dealloc and dispose all observations, some
new API's are required so that we can unplug/replug feedbacks to
achieve the above mentioned pause/resume behavior.

## Changes

- Create new `plugFeedbacks` and `unplugFeedbacks` API's in
`Floodgate`, which establish and dispose feedbacks observations,
respectively. Floodgate now retains the feedbacks passed in on
`bootstrap` to use them on `plugFeedbacks`.

- Add `pause` and `resume` API's to `LoopBoxBase`.

- Implement `pause` and `resume` API's in `RootLoopBox`, which unplug
and plug the feedbacks on the `Floodgate`, respectively.

- Implement `pause` and `resume` API's in `ScopedLoopBox`, which
forward the calls to their root, respectively.

p4checo commented on Loop/Floodgate.swift:
```diff
@@ -40,7 +41,25 @@ final class Floodgate<State, Event>: FeedbackEventConsumer<Event> {
+        }
+    }
+
+    func unplugFeedbacks() {
```
not sure about the naming, so I am open to suggestions 😇

p4checo commented on Loop/Floodgate.swift:
```diff
@@ -40,7 +41,25 @@ final class Floodgate<State, Event>: FeedbackEventConsumer<Event> {
+        feedbackDisposables = CompositeDisposable()
+    }
+
+    func replugFeedbacks() {
```
thought about simply invoking `bootstrap(with:)` again from the `Loop`, but decided that the `Floodgate` should be the one retaining the feedbacks. It also allowed having a symmetrical API (unplug/replug)

However the `Loop` being the one retaining the feedbacks also seems like a valid approach.

EDIT: renamed `replugFeedbacks` to `plugFeedbacks`, and invoke `plugFeedbacks` from `bootstrap` to avoid duplicated code.

p4checo commented on Loop/LoopBox.swift:
```diff
@@ -42,6 +42,14 @@ internal class ScopedLoopBox<RootState, RootEvent, ScopedState, ScopedEvent>: Lo
     }
+
+    override func pause() {
+        root.pause()
```
not sure about this approach, as this way one scoped LoopBox will have the "power" to pause the root (parent), and as a consequence all other sibling scoped loop boxes.

p4checo reviewed (commented):
Some points where I'm not certain about the chosen approach 😇

andersio commented:
Having given some thoughts about this over the weekend, I think making `resume()` and `pause()` available on `Loop` can make it potentially confusing when it comes to semantics around scoped stores (as you briefly mentioned).

I wonder if we would venture a state-driven approach instead. Say given this state definition:

```swift
struct State {
  var messages: [Message] = []
  var isPaused: Bool = false
}
```

Then we can define pausable feedbacks through a feedback modifier, that instructs the feedback to listen to a particular predicate. So Loop users also get to control and program the scope and granularity of cancellation.

```swift
let loop = Loop(
  initial: State(),
  reducer: ...,
  feedbacks: [
    Loop.Feedback
        .combine(
          Self.feedback_pausable_whenLoading(),
          Self.feedback_pausable_whenX(),
          Self.feedback_pausable_whenY(),
          Self.feedback_pausable_whenZ()
        )
        .autoconnect(condition: { $0.isPaused == false })
  ]
)
```

(... while `Loop.send` remains always available.)

P.S. I have a bit of hesitation about _pause_ as the term too, since it usually implies progress preservation which would not be the case for Loop (unless it being explicitly programmed to do so). So I chose the term _autoconnect_ in the snippet above.

p4checo commented:
Sorry for the delay, I was off last week 😇 

I agree that the `resume()` and `pause()` APIs can be a bit confusing for users. I went with it because on initial conversations with @RuiAAPeres he suggested that it would be the preferred direction for this. My initial approach was to simply use `start()` and `stop()` APIs, however they have been deprecated. Out of curiosity: what was the reasoning for it?

On our particular case, the need for this arose because we are using Loop to model Services on our app (e.g. responsible for performing network requests and managing a local state) which we want to start/stop. Even though we currently think of the start/stop mechanics as _external_ to the state machine, I believe that they can (and probably should) be an integral part of it. Following this state-driven approach we keep this behavior nicely packed in the state which makes everything more explicit while also being more extensible and composable. ✨ 

Regarding the term, I agree that _pause_ is not ideal and I think that _autoconnect_ is definitely an improvement, despite personally not being fully convinced by it either 😅. Since I couldn't come up with a better one, I'll go with it for now.

p4checo committed:
Implement `Feedback.autoconnect` operator

- Implement a new `autoconnect` operator  on `Feedback` that disposes
and restores its "parent" Feedback observation according to a predicate
on `State`.

p4checo commented on Loop/Public/FeedbackLoop.swift:
```diff
@@ -453,6 +453,17 @@ extension Loop {
             }
         }
 
+        public func autoconnect(followingChangesIn predicate: @escaping (State) -> Bool) -> Feedback {
+            return Feedback { state, consumer in
+                self.events(
+                    state
+                        .skipRepeats { predicate($0.0) == predicate($1.0) }
+                        .flatMap(.latest) { predicate($0.0) ? state.filter { predicate($0.0) } : .empty },
+                    consumer
+                )
+            }
+        }
+
```
This is what I managed to come up with, and it appears to be working. However I am not convinced this is the best approach... 🙈  

More specifically, I had to add that `filter` on the inner `state` to avoid the "disconnect-triggering" state from being processed before the inner chain is replaced.

Any ideas to make this tidier and/or more idiomatic?
